### PR TITLE
Code Coverage Thresholds and Flags

### PR DESCRIPTION
**Pull request by Dhruv-J**
This PR provides the following updates for code coverage:
1. defining thresholds for a patch (60%)
2. adding kind-e2e-test flags and functionality

#163

### CODE REVIEW

File: codecov.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@ coverage:
   status:
     patch:
       default:
-        target: auto
+        target: 60%
```
Comment (elton-furtado) [inline codecov.yml:21]: you can mark this as 70 after this PR goes in  https://github.com/antrea-io/theia/pull/230

File: codecov.yml [hunk 2]
```diff
@@ -29,12 +29,19 @@ coverage:
...
       theia-unit-tests:
         target: auto
+        threshold: 1%
         flags:
           - unit-tests
+      theia-kind-e2e-tests:
+        target: auto
+        threshold: 1%
+        flags:
+          - kind-e2e-tests
```
Comment (elton-furtado) [inline codecov.yml:40]: e2e-tests section also needs to be added

File: .github/workflows/kind.yml [hunk 3]
```diff
@@ -153,7 +153,15 @@ jobs:
...
-          ANTREA_LOG_DIR=$PWD/log ./ci/kind/test-e2e-kind.sh
+          mkdir .coverage
+          ANTREA_LOG_DIR=$PWD/log ./ci/kind/test-e2e-kind.sh --coverage
+      - name: Codecov
+        uses: codecov/codecov-action@v3
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          file: .coverage/coverage-e2e.txt
+          flags: kind-e2e-tests
+          name: codecov-e2e-encap-test
```
Comment (dreamtalen) [inline .github/workflows/kind.yml:164]: Unlike Antrea, Theia doesn't have `encap` mode.
So I recommend renaming this to `codecov-kind-e2e-test`

**Review by dreamtalen — COMMENTED** [review]
Thank you Dhruv for working on this.
While checking the[ codecov file of kind-e2e-test](https://app.codecov.io/github/antrea-io/theia/commit/2bfbd0b04e023e301b3effac4699874a3c3cd053) uploaded by this PR, I noticed that only files under `test/e2e/` have code coverage lines reported.
In comparison, I checked the Antrea side and their kind-e2e codecov file includes code coverage information for functional golang files as well, [for example, Click the 'Download' button on any item with the 'kind-e2e-test' flag](https://app.codecov.io/gh/antrea-io/antrea/commit/8dfc47aa5d5fd0c269fea98817ae692a7ded13e7).
Could you check if there are any other changes we missing to report the accurate code coverage for kind-e2e tests?

File: .github/workflows/kind.yml [hunk 4]
```diff
@@ -154,7 +154,15 @@ jobs:
       - name: Run e2e tests
         run: |
           mkdir log
-          ANTREA_LOG_DIR=$PWD/log ./ci/kind/test-e2e-kind.sh
+          mkdir .coverage
```
Comment (elton-furtado) [inline .github/workflows/kind.yml:157]: try using test-e2e-converage folder name

File: .github/workflows/kind.yml [hunk 5]
```diff
@@ -154,7 +154,15 @@ jobs:
...
+          ANTREA_LOG_DIR=$PWD/log ./ci/kind/test-e2e-kind.sh --coverage
+      - name: Codecov
```
Comment (elton-furtado) [inline .github/workflows/kind.yml:159]: - name: Tar coverage files
      run: tar -czf test-e2e-coverage.tar.gz test-e2e-coverage
    - name: Upload coverage for test-e2e-coverage
      uses: actions/upload-artifact@v3
      with:
        name: test-e2e-coverage
        path: test-e2e-coverage.tar.gz
        retention-days: 30
        
isn't these steps required as well?

File: .github/workflows/kind.yml [hunk 6]
```diff
@@ -154,7 +154,15 @@ jobs:
...
+        uses: codecov/codecov-action@v3
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          file: .coverage/coverage-e2e.txt
```
Comment (elton-furtado) [inline .github/workflows/kind.yml:163]: file: '*.cov.out*'

File: .github/workflows/kind.yml [hunk 7]
```diff
@@ -154,7 +154,15 @@ jobs:
...
+          flags: kind-e2e-tests
+          name: codecov-kind-e2e-test
```
Comment (elton-furtado) [inline .github/workflows/kind.yml:165]: directory: test-e2e-coverage
        fail_ci_if_error: ${{ github.event_name == 'push' }}

File: ci/kind/test-e2e-kind.sh [hunk 8]
```diff
@@ -172,7 +178,11 @@ function run_test {
   rm -rf $TMP_DIR
   sleep 1
 
-  go test -v -timeout=30m antrea.io/theia/test/e2e -provider=kind --logs-export-dir=$ANTREA_LOG_DIR --skip=$skiplist
+  if $coverage; then
+    go test -v -timeout=30m antrea.io/theia/test/e2e -provider=kind --logs-export-dir=$ANTREA_LOG_DIR -cover -coverpkg=antrea.io/theia/plugins/...,antrea.io/theia/pkg/... -coverprofile=.coverage/coverage-e2e.txt -covermode=atomic --skip=$skiplist
```
Comment (elton-furtado) [inline ci/kind/test-e2e-kind.sh:182]: $RUN_OPT missing?
$coverage_args missing?

File: codecov.yml (hunk 2 above)
Comment (elton-furtado) [reply codecov.yml:41]: antrea has removed the e2e section recently, might want to check that out

File: .github/workflows/kind.yml (hunk 5 above)
Comment (Dhruv-J) [reply .github/workflows/kind.yml:159]: Nope, it seems to work for unit-tests flag fine without it.

File: ci/kind/test-e2e-kind.sh (hunk 8 above)
Comment (Dhruv-J) [reply ci/kind/test-e2e-kind.sh:182]: I don't think we have additional run options or another set of arguments to run coverage with, so might not be needed.

**Comment by Dhruv-J** [comment]
> Thank you Dhruv for working on this. While checking the[ codecov file of kind-e2e-test](https://app.codecov.io/github/antrea-io/theia/commit/2bfbd0b04e023e301b3effac4699874a3c3cd053) uploaded by this PR, I noticed that only files under `test/e2e/` have code coverage lines reported. In comparison, I checked the Antrea side and their kind-e2e codecov file includes code coverage information for functional golang files as well. Could you check if there are any other changes we missing to report the accurate code coverage for kind-e2e tests?

So I did a lot of digging into potential differences in setups for coverage, but I wasn't able to find anything I haven't covered. I think that the current coverage files are all that are listed, and overall coverage is still calculated correctly.

**Comment by dreamtalen** [comment]
> > Thank you Dhruv for working on this. While checking the[ codecov file of kind-e2e-test](https://app.codecov.io/github/antrea-io/theia/commit/2bfbd0b04e023e301b3effac4699874a3c3cd053) uploaded by this PR, I noticed that only files under `test/e2e/` have code coverage lines reported. In comparison, I checked the Antrea side and their kind-e2e codecov file includes code coverage information for functional golang files as well. Could you check if there are any other changes we missing to report the accurate code coverage for kind-e2e tests?
> 
> So I did a lot of digging into potential differences in setups for coverage, but I wasn't able to find anything that could be causing the problem. I went over all the go test flags, directory setup, and changes made in a similar PR for Antrea. I think that the current coverage files are all that are listed, and overall coverage is still calculated correctly.

Hi Dhruv, thanks for looking into it. I checked the latest coverage file again, but I still believe the coverage number is not correct. To be specific, our `kind-e2e-tests` only reports the code coverage under the `/test` folder but not functional golang code under `/pkg` and `/plugins`. 
Here is the evidence in the coverage link of this PR https://app.codecov.io/github/antrea-io/theia/commit/982d79216268d1921d61d4f4a1654e4693707169/tree, which shows a coverage % of 44.39% for the `/test` folder.
image
While in the code coverage comment in this PR, `kind-e2e-tests` flag also has an exact same coverage 44.39%, which implies that `kind-e2e-tests` didn't provide any code coverage info for other folders:
image
For comparison, here is the coverage link of Antrea main branch: https://app.codecov.io/github/antrea-io/antrea
image
We can see that Antrea doesn't report coverage under the `/test` folder and their `kind-e2e-test` has a coverage 44% (link: https://app.codecov.io/github/antrea-io/antrea/flags)
image, which is not the same coverage as the `/test` folder or any other single folder. This is because their `kind-e2e` codecov file includes code coverage information for functional golang files across multiple folders under `/pkg`.

In summary, I would like to clarify two things: 

- Firstly, do we also not need to report the code coverage under the `/test` folder, like Antrea? 
- Secondly, can we modify the `kind-e2e-tests` flag to report only the coverage of functional code, and not test code?

I hope these info could be helpful for futher investigation, Thanks!

**Comment by Dhruv-J** [comment]
> I hope these info could be helpful for futher investigation, Thanks!

Yeah, this is really helpful, thanks!

**Comment by elton-furtado** [comment]
We should add a task to report  all the python test files coverage
